Turbopack: Document new advanced `turbopack.rule.*.condition` config options (#83246)
vercel/next.js: docs/01-app/03-api-reference/05-config/01-next-config-js/turbopack.mdx

> }

Reviewer: <!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'><i>[Re: line +109]</i></p><!--__GRAPHITE_HTML_TAG_END__-->
does a simple rule like this match in node_modules?

there should be a recommendation somewhere about specifying `not: 'foreign'` for performance up here
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.dev/github/pr/vercel/next.js/83246?utm_source=unchanged-line-comment">Graphite</a>.</p><!--__GRAPHITE_HTML_TAG_END__-->

Author: I'm a bit conflicted. I think it's a good idea in general, but I also don't want to add too much complexity up-front.

We had an earlier discussion about if `not: foreign` should be the default. It was pointed out that unless you're matching against `*.js`, you probably won't actually match many/any files in `node_modules` with the glob.

https://vercel.slack.com/archives/C046HAU4H7F/p1755301695633599

You can further restrict where a loader runs using the advanced `condition` syntax:

```js filename="next.config.js"
module.exports = {
  turbopack: {
    rules: {
      // '*' will match all file paths, but we restrict where our
      // rule runs with a condition.
      '*': {
        condition: {
          all: [
            // 'foreign' is a built-in condition.
            { not: 'foreign' },
            // 'path' can be a RegExp or a glob string. A RegExp matches
            // anywhere in the full project-relative file path.
            { path: /^img\/[0-9]{3}\// },
            {
              any: [
                { path: '*.svg' },
                // 'content' is always a RegExp, and can match
                // anywhere in the file.
                { content: /\<svg\W/ },
              ],
            },
          ],
        },
        loaders: ['@svgr/webpack'],
        as: '*.js',
      },
    },
  },
}
```

- Supported boolean operators are `{all: [...]}`, `{any: [...]}` and `{not: ...}`.
- Supported customizable operators are `{path: string | RegExp}` and `{content: RegExp}`. If `path` and `content` are specified in the same object, it acts as an implicit `and`.

> In addition, a number of special built-in conditions are supported:

Reviewer: ```suggestion
In addition, a number of built-in conditions are supported:
```

- `default`: Always matched. A no-op.

> - `browser`: The code will be executed on the client.

Reviewer: instead of `The code will be executed on the client.`  maybe `Matches code that will execute on the client'`... and throughout

- `foreign`: The code is part of `node_modules` or is a Next.js internal. Usually you'll want to restrict loaders to `{not: 'foreign'}`. This can improve performance by reducing the number of files the loader is invoked on.
- `development`: The loader will only be run during `next dev --turbopack`.
- `production`: The loader will only be run during `next build --turbopack`.
- `node`: The code will be run on the default Node.js runtime.
- `edge-light`: The code will be run on the [Edge runtime](/docs/app/api-reference/edge).

> Rules can be an array of objects. All matching rules are executed in order. Array can be useful when used with disjoint conditions:

Reviewer: ```suggestion
Rules can also be an array of objects which is useful for modeling disjoint conditions: 

> **Good to know**: All matching rules are executed in order.
```

Something like that.  i found the phrasing a little awkward

```js filename="next.config.js"
module.exports = {
  turbopack: {
    rules: {
      "*.svg": [
        {
          condition: 'browser',
          loaders: ["@svgr/webpack"],
          as: "*.js",
        },
        {
          condition: { not: 'browser' },
          loaders: [require.resolve('./custom-svg-loader.js')],

>           as: "*.js",

Reviewer 2: I think the lint might fail here - should be `'` single quotes - same in 199, 202, 203

Author: Thanks, I plugged this into prettier at one point to double-check my work and accidentally changed a bunch of stuff to double quotes.

```js filename="next.config.js"
module.exports = {
  turbopack: {
    rules: {
      // '*' will match all file paths, but we restrict where our
      // rule runs with a condition.
      '*': {
        condition: {
          all: [
            // 'foreign' is a built-in condition.
            { not: 'foreign' },
            // 'path' can be a RegExp or a glob string. A RegExp matches
            // anywhere in the full project-relative file path.
            { path: /^img\/[0-9]{3}\// },
            {
              any: [
                { path: '*.svg' },
                // 'content' is always a RegExp, and can match
                // anywhere in the file.
                { content: /\<svg\W/ },
              ],
            },
          ],
        },
        loaders: ['@svgr/webpack'],
        as: '*.js',
      },
    },
  },
}
```

- Supported boolean operators are `{all: [...]}`, `{any: [...]}` and `{not: ...}`.
- Supported customizable operators are `{path: string | RegExp}` and `{content: RegExp}`. If `path` and `content` are specified in the same object, it acts as an implicit `and`.

In addition, a number of built-in conditions are supported:

> - `default`: Always matched. A no-op.

Reviewer: i guess this is useful for composing with other conditions

`process.env.DEVELOPMENT == "1" ? some-condition : 'default'` 

?

Author: Yeah, but there's other ways to handle that.

This condition seems like it might be a holdover from the old builtin condition syntax where it might've been more useful: https://www.notion.so/vercel/Turbopack-loader-rule-syntax-254e06b059c4809096d1e7c9afad278c?source=copy_link#254e06b059c4803a93a2f3199c77ab5d

I can remove support for this if you think it's a good idea.

Author: Follow up PR: https://github.com/vercel/next.js/pull/84111
